src/Bootstrappers/PrefixCacheTenancyBootstrapper.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Bootstrappers;
+
+use Illuminate\Cache\Repository as CacheRepository;
+use Illuminate\Contracts\Config\Repository;
+use Illuminate\Contracts\Foundation\Application;
+use Illuminate\Support\Facades\Cache;
+use Stancl\Tenancy\Contracts\TenancyBootstrapper;
+use Stancl\Tenancy\Contracts\Tenant;
+
+class PrefixCacheTenancyBootstrapper implements TenancyBootstrapper
+{
+    protected null|string $originalPrefix = null;
+    protected string $storeName;
+
+    public function __construct(
+        protected Application $app,
+        protected Repository $config,
+    ) {
+    }
+
+    public function bootstrap(Tenant $tenant): void
+    {
+        $this->originalPrefix = $this->config->get('cache.prefix');
+        $this->storeName = $this->config->get('cache.default');
+
+        $this->setCachePrefix($this->config->get('tenancy.cache.prefix_base') . $tenant->getTenantKey());
+    }
+
+    public function revert(): void
+    {
+        $this->setCachePrefix($this->originalPrefix);
+        $this->originalPrefix = null;
+    }
+
+    protected function setCachePrefix(null|string $prefix): void
+    {
+        $this->config->set('cache.prefix', $prefix);
+
+        $this->app['cache']->forgetDriver($this->storeName);
+
+        // The CacheManager will have the $app['config'] array cached with old prefixes on the 'cache' instance
+        // This call will forget the 'cache' instance
+        $this->app->forgetInstance('cache');
```

Comment:
> Is the driver stored as a singleton instance in the container or something like that?

Yes. https://github.com/laravel/framework/blob/9.x/src/Illuminate/Cache/CacheServiceProvider.php#L22

Reply:
I don't think that's related to `forgetDriver()` — we remove `cache.store` on line 50, so what exactly does `forgetDriver()` do here?

Reply:
Seems like the method itself only calls `unset()` **on the `CacheManager`**, so it doesn't interact with the container. Therefore this is false:

> Is the driver stored as a singleton instance in the container or something like that?

And if we forget the instance of `CacheManager` completely on line 47, I don't think we need to bother cleaning up the state of the class?

Or is this due to existing classes that may already have the `CacheManager` injected?

Reply:
> so it doesn't interact with the container. Therefore this is false:

Ah I see. 

> And if we forget the instance of CacheManager completely on line 47, I don't think we need to bother cleaning up the state of the class?

Logically it makes sense. I tried removing it tests are still passing.

Reply:
Presumably because we aren't testing a case where `CacheManager` is injected into a service class, and then tenancy ended and reinitialized for another tenant.

Reply:
What are you suggesting here? Should we write a test?

Reply:
Probably yeah. And check if that verifies that the `forgetDriver()` call here is used for that.

Reply:
@stancl I wrote this test https://github.com/archtechx/tenancy/pull/1014/commits/222686ec1d6772048ded1b6de8e73fac96347bbf, and it looks like `CacheManager` injected into a service class working fine without `$this->app['cache']->forgetDriver($this->storeName);`

Reply:
- Try to use a second tenant as well, as I mentioned here
- The service should be a singleton. You're creating a fresh instance every time, which misses the point of what we want to test. We want to have a single instance of the service and then see if its dependencies **that were already injected in the past** are updated. Here you're re-creating the instance and re-injecting a fresh `CacheManager`

Reply:
> The service should be a singleton. 

That is a very good point. The test fails when I register it as a singleton. So I enabled this line `$this->app['cache']->forgetDriver($this->storeName);` and the test is passing.

Reply:
Looks like it's failing for `tenant2`. Looking into it.

Reply:
I think we can resolve this because the code is different now https://github.com/archtechx/tenancy/pull/1014/files#diff-41b4083e1eb3004d0f552f7805c28a2c2e479078ed592be5d955a90788f486d5R41